cmd/renterd/node.go

```diff
@@ -0,0 +1,419 @@
+			as.V4Keypairs = make(map[string]string)
+		}
+
+		// S3 key pair validation was broken at one point, we need to remove the
+		// invalid key pairs here to ensure we don't fail when we update the
+		// setting below.
+		for k, v := range as.V4Keypairs {
+			if err := (api.S3AuthenticationSettings{V4Keypairs: map[string]string{k: v}}).Validate(); err != nil {
+				n.logger.Infof("removing invalid S3 keypair for AccessKeyID %s, reason: %v", k, err)
+				delete(as.V4Keypairs, k)
+			}
+		}
+
+		// merge keys
+		for k, v := range n.cfg.S3.KeypairsV4 {
+			as.V4Keypairs[k] = v
+		}
+		// update settings
+		if err := n.bus.UpdateSetting(context.Background(), api.SettingS3Authentication, as); err != nil {
+			return fmt.Errorf("failed to update S3 authentication settings: %w", err)
+		}
+	}
+
+	// start S3 server
+	if n.s3Srv != nil {
+		go n.s3Srv.Serve(n.s3Listener)
+		n.logger.Info("s3: Listening on " + n.s3Listener.Addr().String())
+	}
+
+	// fetch the syncer address
+	syncerAddress, err := n.bus.SyncerAddress(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to fetch syncer address: %w", err)
+	}
+	n.logger.Info("bus: Listening on " + syncerAddress)
+
+	// run autopilot store migration
+	//
+	// TODO: we can safely remove this already
+	if n.cfg.Autopilot.Enabled {
+		autopilotDir := filepath.Join(n.cfg.Directory, n.cfg.Autopilot.ID)
+		if err := runCompatMigrateAutopilotJSONToStore(n.bus, "autopilot", autopilotDir); err != nil {
+			return fmt.Errorf("failed to migrate autopilot JSON: %w", err)
+		}
+	}
+
+	// open the web UI if enabled
+	if n.cfg.AutoOpenWebUI {
+		time.Sleep(time.Millisecond) // give the web server a chance to start
+		_, port, err := net.SplitHostPort(n.apiListener.Addr().String())
+		if err != nil {
+			n.logger.Debug("failed to parse API address", zap.Error(err))
+		} else if err := utils.OpenBrowser(fmt.Sprintf("http://127.0.0.1:%s", port)); err != nil {
+			n.logger.Debug("failed to open browser", zap.Error(err))
+		}
+	}
+	return nil
+}
+
+func (n *node) Shutdown() error {
+	n.logger.Info("Shutting down...")
+
+	// give each service a fraction of the total shutdown timeout. One service
+	// timing out shouldn't prevent the others from attempting a shutdown.
+	timeout := n.cfg.ShutdownTimeout / time.Duration(len(n.shutdownFns))
+	shutdown := func(fn func(ctx context.Context) error) error {
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+		return fn(ctx)
+	}
+
+	// shut down the services in reverse order
+	var errs []error
+	for i := len(n.shutdownFns) - 1; i >= 0; i-- {
+		if err := shutdown(n.shutdownFns[i].fn); err != nil {
+			n.logger.Errorf("failed to shut down %v: %v", n.shutdownFns[i].name, err)
+			errs = append(errs, err)
+		} else {
+			n.logger.Infof("%v shut down successfully", n.shutdownFns[i].name)
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+func runCompatMigrateAutopilotJSONToStore(bc *bus.Client, id, dir string) (err error) {
```

Comment:
@ChrisSchinnerl I'd remove this in a F/U - it's been there since before 1.0 I believe so we can safely get rid of it.